tests/platform_tests/api/test_psu.py

```diff
@@ -206,66 +207,75 @@ def test_power(self, duthosts, enum_rand_one_per_hwsku_hostname, localhost, plat
+                                        "PSU {} maximum supplied power appears incorrect".format(psu_id))
+
+                    if current is not None and voltage is not None and power is not None:
+                        is_within_tolerance = abs(power - (voltage*current)) < power*0.1
+                        if not is_within_tolerance and i < MAX_ATTEMPTS - 1:
+                            logger.info("Retrying test for {} as readings not within tolerance level".format(name))
+                            continue
+                        self.expect(is_within_tolerance, "PSU {} reading does not make sense \
+                            (power:{}, voltage:{}, current:{})".format(psu_id, power, voltage, current))
+
+                    powergood_status = psu.get_powergood_status(platform_api_conn, psu_id)
+                    if self.expect(powergood_status is not None,
+                                   "Failed to retrieve operational status of PSU {}".format(psu_id)):
+                        self.expect(powergood_status is True, "PSU {} is not operational".format(psu_id))
+
+                    high_threshold = None
+                    voltage_high_threshold_supported = self.get_psu_facts(duthost, psu_id, True,
+                                                                          "voltage_high_threshold")
+                    if voltage_high_threshold_supported:
+                        high_threshold = psu.get_voltage_high_threshold(platform_api_conn, psu_id)
+                        if self.expect(high_threshold is not None,
+                                       "Failed to retrieve the high voltage threshold of PSU {}".format(psu_id)):
+                            self.expect(isinstance(high_threshold, float),
+                                        "PSU {} voltage high threshold appears incorrect".format(psu_id))
+                    low_threshold = None
+                    voltage_low_threshold_supported = self.get_psu_facts(duthost, psu_id, True, "voltage_low_threshold")
+                    if voltage_low_threshold_supported:
+                        low_threshold = psu.get_voltage_low_threshold(platform_api_conn, psu_id)
+                        if self.expect(low_threshold is not None,
+                                       "Failed to retrieve the low voltage threshold of PSU {}".format(psu_id)):
+                            self.expect(isinstance(low_threshold, float),
+                                        "PSU {} voltage low threshold appears incorrect".format(psu_id))
+                    if high_threshold is not None and low_threshold is not None:
+                        self.expect(voltage < high_threshold and voltage > low_threshold,
+                                    "Voltage {} of PSU {} is not in between {} and {}"
+                                    .format(voltage, psu_id, low_threshold, high_threshold))
+
+                break
+
         self.assert_expectations()
```

Comment:
How does the self.assert_expectations() evaluate the 3 test run values? What happens for different combinations of the expected values? For example:

Test1: True
Test2: False
Test3: True

In this combination, does the test raise assert or passes?

Reply:
The test would fail in this scenario. I have an update to address it, will commit the changes. Thanks.

Reply:
@smagarwal-arista With new changes, does the test pass in above scenario? Also, does the test exit on first Passed case?

Reply:
Yes, the test would exit the retry loop if all expects pass.  Retry is only targeted for expect failing on `abs(power - (voltage*current)) < power*0.1` condition. Any other expect failures would exit the retry loop for that PSU and eventually result in a failed test.